Introduce new 'import-private-name' checker

areveny opened the pull request:
## Type of Changes


|     | Type                   |
| --- | ---------------------- |
| ✓   | :sparkles: New feature |

## Description

Hi folks, 

I used one convention message  for modules and objects, customizing the message to help identify the type of the private name being imported, i.e.
```
import-private-name:4:0:4:25::Imported private package (_world):UNDEFINED
import-private-name:5:0:5:23::Imported private object (_house):UNDEFINED
import-private-name:6:0:6:29::Imported private objects (_b, _d):UNDEFINED
```

Right now it's just a simple check for the name leading with a single underscore `_`. 

The results of running this check on Pylint itself are below with single and double underscore cases. 

```
************* Module checkers.unittest_misc
tests/checkers/unittest_misc.py:22:0: C0416: Imported private object (_tokenize_str) (import-private-name)
************* Module pylint.testutils.functional.lint_module_output_update
pylint/testutils/functional/lint_module_output_update.py:8:0: C0416: Imported private package (_pytest.config) (import-private-name)
************* Module pylint.config.option_manager_mixin
pylint/config/option_manager_mixin.py:20:0: C0416: Imported private object (_ManHelpFormatter) (import-private-name)
************* Module checkers.unittest_format
tests/checkers/unittest_format.py:39:0: C0416: Imported private object (_tokenize_str) (import-private-name)
************* Module checkers.unittest_spelling
tests/checkers/unittest_spelling.py:27:0: C0416: Imported private object (_tokenize_str) (import-private-name)
************* Module pylint.lint.run
pylint/lint/run.py:8:0: C0416: Imported private object (__pkginfo__) (import-private-name)
pylint/lint/run.py:9:0: C0416: Imported private object (_config_initialization) (import-private-name)
************* Module test_check_parallel
tests/test_check_parallel.py:25:0: C0416: Imported private object (_worker_check_single_file) (import-private-name)
tests/test_check_parallel.py:26:0: C0416: Imported private object (_worker_initialize) (import-private-name)
************* Module pylint.config
pylint/config/__init__.py:48:0: C0416: Imported private object (_ManHelpFormatter) (import-private-name)
************* Module pylint
pylint/__init__.py:15:0: C0416: Imported private object (__version__) (import-private-name)
************* Module test_functional
tests/test_functional.py:30:0: C0416: Imported private package (_pytest.config) (import-private-name)
tests/test_functional.py:31:0: C0416: Imported private package (_pytest.recwarn) (import-private-name)
************* Module pylint.checkers.utils
pylint/checkers/utils.py:83:0: C0416: Imported private module (_string) (import-private-name)
************* Module test_import_graph
tests/test_import_graph.py:27:0: C0416: Imported private package (_pytest.fixtures) (import-private-name)
************* Module pylint.extensions.docparams
pylint/extensions/docparams.py:37:0: C0416: Imported private object (_check_docs_utils) (import-private-name)
************* Module lint.unittest_expand_modules
tests/lint/unittest_expand_modules.py:12:0: C0416: Imported private object (_is_in_ignore_list_re) (import-private-name)
************* Module pylint.lint.parallel
pylint/lint/parallel.py:9:0: C0416: Imported private object (_patch_sys_path) (import-private-name)
************* Module pylint.testutils
pylint/testutils/__init__.py:53:0: C0416: Imported private object (_get_tests_info) (import-private-name)
pylint/testutils/__init__.py:58:0: C0416: Imported private object (_tokenize_str) (import-private-name)
************* Module pylint.config.options_provider_mixin
pylint/config/options_provider_mixin.py:8:0: C0416: Imported private object (_validate) (import-private-name)
************* Module pylint.constants
pylint/constants.py:9:0: C0416: Imported private object (__version__) (import-private-name)
************* Module pylint.checkers.base_checker
pylint/checkers/base_checker.py:27:0: C0416: Imported private object (_MSG_ORDER) (import-private-name)
************* Module pylint.lint
pylint/lint/__init__.py:84:0: C0416: Imported private object (_patch_sys_path) (import-private-name)
************* Module test_func
tests/test_func.py:30:0: C0416: Imported private object (_get_tests_info) (import-private-name)
************* Module pylint.reporters.text
pylint/reporters/text.py:48:0: C0416: Imported private object (_splitstrip) (import-private-name)
************* Module pylint.extensions.overlapping_exceptions
pylint/extensions/overlapping_exceptions.py:13:0: C0416: Imported private object (_annotated_unpack_infer) (import-private-name)
************* Module pylint.utils
pylint/utils/__init__.py:51:0: C0416: Imported private objects (_check_csv, _format_option_value, _splitstrip, _unquote) (import-private-name)
************* Module pylint.message.message_definition
pylint/message/message_definition.py:9:0: C0416: Imported private object (_SCOPE_EXEMPT) (import-private-name)
************* Module extensions.test_check_docs_utils
tests/extensions/test_check_docs_utils.py:21:0: C0416: Imported private object (_check_docs_utils) (import-private-name)
************* Module pylint.testutils.lint_module_test
pylint/testutils/lint_module_test.py:15:0: C0416: Imported private package (_pytest.config) (import-private-name)
pylint/testutils/lint_module_test.py:18:0: C0416: Imported private object (_config_initialization) (import-private-name)
pylint/testutils/lint_module_test.py:21:0: C0416: Imported private objects (_EXPECTED_RE, _OPERATORS) (import-private-name)
```

Closes #5463

areveny commented:
I have some test errors to repair but wanted to open this PR to ask how we should handle the existing private imports in the Pylint code. Thanks.

Pierre-Sassoulas commented:
First thank you for handling yet another checker !

> how we should handle the existing private imports in the Pylint code

I think we should add the ckeck to pylint (= not disable it everywhere indefinitely). But considering the number of problems detected we might want to handle this in a latter MR (add the check with a global disable first, fix later). The way we fix it will depends, we can disable the warning or rename the structure we were importing. Sometime, the intent was to inform plugin developer to not rely on a structure we consider private. We might have to specify what exactly is public API if we rename some of those. I expect to be some discussion around the topic when resolving this in the next MR.

Regarding the checker itself, I will do a full review later, but I think ``__version__`` or ``__pkginfo__`` is not protected/private and should be treated as a special case.

areveny commented:
Thanks! I added a commit to fix an external test error and disable check from the built-in `pylintrc`.

Some  integration tests are failing because they now emit `import-private-name`. Some of these cases are `import __future__` so I will wait for your decision on these before adding `import-private-name` to the `# pylint: disable` of these tests.

DanielNoord commented:
I think some of the warnings here show a common case of importing private names: internal use or internal testing.
I would need to recheck PEP8 (and possibly other), but isn't internal use (as in another module within the same package) of private names "allowed"? If not, how should you name a function that can be used within the package but isn't part of public API (for example `utils` functions)?
Another issue shown by the warning messages is importing such names for testing. Although you often want to check the functions that implement these private names, I can see cases where it makes sense to test them specifically (for example an `utils` function again).

I can think of some solutions (allowing internal imports, allowing internal imports when `pytest/unittest` is imported) but I think these "false positives" warrant a discussion whether this should be an extension (or a check at all).

areveny commented:
I think there is pretty good support in [PEP8](https://www.python.org/dev/peps/pep-0008/#naming-conventions) for this at least as an extension. (There is also guidance on `__` but it is more technical.)

Descriptive: Naming Styles:
> _single_leading_underscore: weak "internal use" indicator. E.g. from M import * does not import objects whose names start with an underscore.

Global Variable Names: 
> Modules that are designed for use via from M import * should use the `__all__` mechanism to prevent exporting globals, or use the older convention of prefixing such globals with an underscore (which you might want to do to indicate these globals are "module non-public").

Method Names and Instance Variables:
> Use one leading underscore only for non-public methods and instance variables.

Designing for Inheritance: 
> Public attributes should have no leading underscores.

Public and Internal Interfaces: 
> Even with `__all__` set appropriately, internal interfaces (packages, modules, classes, functions, attributes or other names) should still be prefixed with a single leading underscore.

But it doesn't discuss your good distinction of internal/external imports. Maybe checking on all imports or just external imports could be a toggle?

DanielNoord commented:
Based on the PEP sections you quoted I don't think this should be a toggle but standard behaviour. Especially the last quote tells use to use leading underscores for internal "stuff". Having this checker check internal imports would thus result in suggestions/warnings that do not conform to PEP8, which I don't think we should want.

This checker for external imports makes more sense, but can create troubles with imports for typing purposes. I think some tool warns about this without considering such imports being used for typing which has annoyed me in the past. Adding a check for whether the import is within `if TYPE_CHECKING` could work, but there are cases where I import for typing without it.
Thus, personally I would always turn this checker off and if we want to go ahead with this it should probably be an extension seeing the likely amount of "false-positives" this can create.

It is annoying though as I do believe it is useful to tell users "Hey, you do know that this is considered a private name right?", but I don't think that warrants requiring people to add `disables` to their `pylintrc` or code when they want to respond "Yes, I know but I want to".

Pierre-Sassoulas commented:
>  can create troubles with imports for typing purposes

If there's structure you must give to a lib when using it and the typing is not public then there's an issue in the lib that should make it public ? (So in this case it's probably a local disable to add ?)

I think that we should limit this checker to external import. But how do we check that this is a test file and that external is actually internal ? File name starting with ``test_`` ? import pytest or unittest ? contain function named test_... with assert in it ? This seems rather fragile.

areveny commented:
I think I prefer an extension that doesn't distinguish between external/internal right now. 

It would be simplest and if a projects wants to allow internal private imports while using the extension to check for external private imports, they can use ``disables``. 

A project that does allow importing names for testing in general might still want to be notified when it's being done to ensure that it's a deliberate decision to not expose something to a public API and still use it externally.

DanielNoord commented:
> > can create troubles with imports for typing purposes
> 
> If there's structure you must give to a lib when using it and the typing is not public then there's an issue in the lib that should make it public ? (So in this case it's probably a local disable to add ?)
> 
> I think that we should limit this checker to external import. But how do we check that this is a test file and that external is actually internal ? File name starting with `test_` ? import pytest or unittest ? contain function named test_... with assert in it ? This seems rather fragile.

I know that for example in the `lxml` package the actual `etree._Element` which represents a node in a xml file is private. I find that I often import it for typing in helper functions that interact with xml's nodes.


> I think I prefer an extension that doesn't distinguish between external/internal right now.

As I said previously I would "vote" against this. That would actively ignore a recommendation from PEP8.

> It would be simplest and if a projects wants to allow internal private imports while using the extension to check for external private imports, they can use `disables`.

There are multiple issues and discussions about this but in general I think we should try to strive for false negatives over false positives: ie. not raise anything at all if we know something is likely to generate many false positives.

areveny commented:
Apologies @DanielNoord, I misread your comment as "internal" meaning "within this file" versus you meaning "internal" as "within this module" which completely flipped my interpretation of it.

I think I understand. I'll work on whatever direction we come up with but agree there are implementation difficulties with determining external imports, and even with those there seem to be many false positives in practice.

DanielNoord commented:
> Apologies @DanielNoord, I misread your comment as "internal" meaning "within this file" versus you meaning "internal" as "within this module" which completely flipped my interpretation of it.
> 
> I think I understand. I'll work on whatever direction we come up with but agree there are implementation difficulties with determining external imports, and even with those there seem to be many false positives in practice.

Ah, that's my bad as well. I wasn't specific enough as well.

To progress this further I propose to take this into the following direction:
- Move checker to an extension
- Find way to exclude imports within the same module/package as those are valid uses of "internal" private names
- (Find way to exclude tests, although those should probably be covered by previous point)
- If possible, exclude imports only used for typing

Without being too bold, for me point is non-negotiable as it contradicts PEP8. If we can't find a way to do this I'd say we close the PR (and perhaps even the issue citing the difficulty we face with actual implementation). Obviously, nothing is set in stone so I'm open to any counter-arguments 😄

Pierre-Sassoulas commented:
The way I see it will always be hard to tell if we're in test for the package, or test for another package. But even if we detect tests file properly this would create false negatives for every other project. Are we in ``pylint/tests/`` (the leading ``pylint`` being the git repository name, it could be renamed) or in another project's test that is using pylint's internal ? Is this an acceptable number of false positives, do we want to maintain a piece of code that determine if a file is a test file, then still have false negatives ? I think we might have to accept explicit disable when testing a "private" interface in order to not deal with fragile 'we're in a test directory' detection method. (If you can't test your private function through the public API then the public API isn't that great ?)

The other points seems doable (from the same namespace = exclude, used later as typing = exclude, is actually a dunder like ``__version__`` = exclude)

areveny commented:
Hi folks, I would like to confirm my approach before a more formal review.

Moved to an extension and checking dunders is straightforward. 

I used the typing guard check to check for the typing, but I'm not sure this fulfills the feedback. We want to allow `import _PrivateClass`` if later in the code we see ``def method(self, class: PrivateClass)``? Reusing the logic for `unused-import` perhaps.

Finally, to distinguish internal/external, I used `isort` like in the imports checker. Two concerns:

One: we need the options `known-standard-library` and `known-third-party` from `imports.py`, is there a standard way to pull these values? Redefining in the extension was giving a conflict.

Second: it would be nice to reuse calling `isort_driver.place_module(module)` in `imports.py` to remove a little overhead. Is there a way to inject extension code and use values from the base checker?

Much appreciated.

Pierre-Sassoulas commented:
> I used the typing guard check to check for the typing, but I'm not sure this fulfills the feedback. We want to allow import _PrivateClass if later in the code we see def method(self, class: _PrivateClass) ? Reusing the logic for unused-import perhaps.

That's clever, I haven't though of this possibility. I thought `` def method(self, class: _PrivateClass)`` should be ok. Do want to force the use of typing guard for private class used only for typing ? I.e. "if you need to import this private class for typing use ``if typing.TYPE_CHECKING:`` and the expected code would be:
```python
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    import _PrivateClass


 def method(self, class: "_PrivateClass")
     pass
```


This would be a lot of info in the same message though. 



> Finally, to distinguish internal/external, I used isort like in the imports checker.

What about using the name of the current module ? Say we're analyzing ``pylint/checkers/foo.py`` and it's importing ``pylint.bar._ProtectedClass`` then it's internal because there's a common namespace (pylint), if we're importing ``astroid.bar._ProtectedClass`` then it's external as there is no shared namespace ?

Using isort's configuration is a double edged swords, it's taking the wish of the user into account because pylint would say ``astroid.bar._ProtectedClass`` is ok if astroid is added as ``known-first-party``. The user probably know what they're doing when configuring isort. But it's also a false positive I think. astroid is not the same namespace than pylint, if we want to import protected things from astroid we should make it public. Also the code we'll make to do that will be a lot more complex to maintain for us (isort integration, whe nwe don't control isort, reading a configuration...). I think it's counter-intuitive for users too : We're pylint, not isort, it's unexpected that we change our behavior depending on the isort config for things not related to import order ? Let me know what you think.

DanielNoord commented:
> I used the typing guard check to check for the typing, but I'm not sure this fulfills the feedback. We want to allow `import _PrivateClass` if later in the code we see `def method(self, class: _PrivateClass)` ? Reusing the logic for `unused-import` perhaps.

Haven't looked at the code or the `isort` thing as @Pierre-Sassoulas responded quite in depth to it.
However, thinking of this: can we make a `util` here? Something like `is_only_used_as_annotation` with:
```python
for reference in nodes.Module.locals[node_name]:
    if isinstance(reference.parent, nodes.AnnAssign) and reference == reference.parent.annotation:
        pass
    else:
        return False
return True
```
Or something with an `all/any` statement. Would that work? I'm not sure actually, but seems like it could be worth a try (also because we can then remove the `if TYPE_GUARDING` check.

areveny commented:
> What about using the name of the current module ?

A new commit  presents two options for filtering internal modules without `isort`.

`same_root_dir` compares the module root name with each directory name in the path of the linted file.

`same_root_package` does the same but checks each directory for an `__init__.py` file to ensure we are only looking at valid Python packages.

Both of these options produce the same output when we run pylint on pylint. Looking a lot more reasonable!
```
************* Module pylint.testutils.lint_module_test
pylint/testutils/lint_module_test.py:15:0: C0416: Imported private module (_pytest.config) (import-private-name)
************* Module pylint.testutils.functional.lint_module_output_update
pylint/testutils/functional/lint_module_output_update.py:8:0: C0416: Imported private module (_pytest.config) (import-private-name)
************* Module test_functional
tests/test_functional.py:30:0: C0416: Imported private module (_pytest.config) (import-private-name)
tests/test_functional.py:31:0: C0416: Imported private module (_pytest.recwarn) (import-private-name)
************* Module pylint.checkers.utils
pylint/checkers/utils.py:83:0: C0416: Imported private module (_string) (import-private-name)
************* Module test_import_graph
tests/test_import_graph.py:27:0: C0416: Imported private module (_pytest.fixtures) (import-private-name)
```

One example difference:
The functional tests fail with `same_root_package` but not with `same_root_dir`. The only directory in the path `/pylint/tests/functional/ext/private_import/private_import.py` with an `__init__.py` is `functional`. (The actual package is in `/pylint/pylint/__init__.py`. )

With `same_root_dir`, it matches the directory name `pylint` with the module name `pylint` even though it's comparing apples to oranges. You can imagine some false negatives from this behavior but here it suppressed a message we don't want in practice.

Either way, I will add unit tests with mocks because these use the directory of the linted file.

On type checking:

> `if isinstance(reference.parent, nodes.AnnAssign) and reference == reference.parent.annotation:`

```
a: SpecialClass

def process_special_class(x: SpecialClass):
    x
    return
```

I'm seeing that we do get `a` in `locals` with type annotation, but in Astroid I can't find the type annotation of `x`.

Pierre-Sassoulas commented:
This is starting to look good, I don't think there's false positive in the new output (if we consider we should use typing guard).

> same_root_package does the same but checks each directory for an __init__.py file to ensure we are only looking at valid Python packages.

The needs for ``__init__.py`` was removed in python 3.3, (a directory without ``__init__.py`` could be a valid implicit namespace package), so I'd rather we use the other algorithm. We could probably create an implicit namespace package where the second algorithm does not work. See [this blog](http://python-notes.curiousefficiency.org/en/latest/python_concepts/import_traps.html#the-missing-init-py-trap)

Regarding getting the typing with astroid, with your example you can do:
```python
from astroid import parse

a = parse("""
a: SpecialClass

def process_special_class(x: SpecialClass):
    x
    return
""")
print(a["a"].parent.annotation.name) # SpecialClass
print(a["process_special_class"]["x"].parent.annotations[0].name) # SpecialClass
```

Pierre-Sassoulas commented on doc/whatsnew/2.13.rst:
```diff
@@ -19,6 +20,10 @@ Removed checkers
 
 Extensions
 ==========
+* New extension ``import-private-name``: indicate imports of external packages and objects
+  that are prefixed with ``_`` and likely to be private.
+
+  Closes #5463
```
```suggestion
* New extension ``import-private-name``: indicate imports of external private packages
  and objects (prefixed with ``_``).

  Closes #5463
```

Pierre-Sassoulas commented on ChangeLog:
```diff
@@ -9,6 +9,11 @@ Release date: TBA
 ..
   Put new features here and also in 'doc/whatsnew/2.13.rst'
 
+* Add new optional  extension ``import-private-name``: indicate imports of external packages and objects
+  that are prefixed with ``_`` and likely to be private.
+
+  Closes #5463
```
```suggestion
* New extension ``import-private-name``: indicate imports of external private packages
  and objects (prefixed with ``_``).

  Closes #5463
```

Pierre-Sassoulas commented on pylint/extensions/private_import.py:
```diff
@@ -0,0 +1,118 @@
+                return True
+
+        private_names = [name for name in names if self._name_is_private(name)]
+        if len(private_names) > 0:
+            private_imports = ", ".join(private_names)
+            self.add_message(
+                "import-private-name",
+                node=node,
+                args=("object" if checking_objects else "module", private_imports),
+            )
+            return True
+        return False
```
```suggestion
        if not private_names:
            return False
        private_imports = ", ".join(private_names)
        type_ = "object" if checking_objects else "module"
        self.add_message("import-private-name", node=node, args=(type_, private_imports))
        return True
```

Pierre-Sassoulas commented on pylint/extensions/private_import.py:
```diff
@@ -0,0 +1,118 @@
+        if not checking_objects:
+            # Check only external modules; internal private imports are allowed
+            names = [name for name in names if not self.same_root_dir(node, name)]
+            if len(names) == 0:
+                return True
```
```suggestion
            if not names:
                return True
```

Pierre-Sassoulas commented on pylint/extensions/private_import.py:
```diff
@@ -0,0 +1,118 @@
+        """Checks if the import is private. Checking an object if checking_objects is True else a module
+        Returns True if a message was emitted
+        """
+        if is_node_in_typing_guarded_import_block(node):
+            return True
```
```suggestion
```

Pierre-Sassoulas commented on pylint/extensions/private_import.py:
```diff
@@ -0,0 +1,118 @@
+        )
+
+    @check_messages("import-private-name")
+    def visit_importfrom(self, node: nodes.ImportFrom) -> None:
```
```suggestion
    def visit_importfrom(self, node: nodes.ImportFrom) -> None:
        if is_node_in_typing_guarded_import_block(node):
            return
```
That way we do the check only once if ``not private_module`` is true.

Pierre-Sassoulas commented on pylint/extensions/private_import.py:
```diff
@@ -0,0 +1,118 @@
+        return False
+
+    @staticmethod
+    def same_root_dir(node: nodes.Import, import_mod_name: str):
```
```suggestion
    def same_root_dir(node: nodes.Import, import_mod_name: str) -> bool:
```

Pierre-Sassoulas commented on pylint/extensions/private_import.py:
```diff
@@ -0,0 +1,118 @@
+        return False
+
+    @staticmethod
+    def same_root_package(node: nodes.Import, import_mod_name: str):
```
```suggestion
    def same_root_package(node: nodes.Import, import_mod_name: str) -> bool:
```

Pierre-Sassoulas commented on pylint/extensions/private_import.py:
```diff
@@ -0,0 +1,118 @@
+        return base_import_package == last_package
+
+    @staticmethod
+    def _name_is_private(name):
```
```suggestion
    def _name_is_private(name: str) -> bool:
```

Pierre-Sassoulas reviewed (commented):
Some small comments, I did not have the time to review everything, sorry

areveny commented:
I incorporated the feedback for the internal/external distinction.

On the type check, once per file, we now save any names used for type checking in a set. If a private import is used as a type annotation, we no longer emit a message.

This is neat but it is quite big, with potential edge cases. I'm also not sure about the performance implications of building this set on every file with any private import. Thanks.

Pierre-Sassoulas commented:
> I'm also not sure about the performance implications of building this set on every file with any private import.

You could do a sanity check using ``python3 -m cProfile -m pylint .pylint_primer_tests/django|grep _populate_type_annotations``. You will need to have launched django's primer at least once with ``pytest -m primer_external_batch_one --primer-external -n auto``. Don't worry if it takes forever, django is huge and this was already slow before your change 😄 There's also smaller project in the primer, like black. For more detailed analysis there's [a new documentation about performance in the doc](https://pylint.pycqa.org/en/latest/development_guide/profiling.html?highlight=profile#performance-analysis-for-pylint)

areveny commented:
Smaller changes:

* Added a try/except on AttributeErrors while checking for type annotations so we don't crash on missing attributes. There was a strange case of this on the stdlib tests for Python 3.10 only I couldn't reproduce locally. Since this is a catch-all it's falling outside of coverage
* Changed the message id to `C2401`
* Functional test coverage of names used in Slices (e.g. Optional[List[TypeA]])

Just remembered I still need to add unit tests with mocking on the internal/external package checker. All the feedback given so far should be incorporated.

DanielNoord commented:
> * Added a try/except on AttributeErrors while checking for type annotations so we don't crash on missing attributes. There was a strange case of this on the stdlib tests for Python 3.10 only I couldn't reproduce locally. Since this is a catch-all it's falling outside of coverage

Do you have logs for this? I'd like us to fix this in this PR as well and I'm happy to take a look to see if I can figure it out!

> * Changed the message id to `C2401`

I know there are two other checkers lined up which use `24` and `25`. To avoid future merge conflicts, could you change this to `2601`? The new script doesn't really work well when there are multiple checkers WIP 😄

areveny commented:
Bumped up the message id. [Here](https://github.com/PyCQA/pylint/runs/4696690018) is a link to one of the failed runs on stdlib in `logging.py`.

I just put in some logic that should guard against this specific crash. Couldn't spot it earlier.

```
        if node.args and hasattr(node.args, "args"):
            for arg in node.args.args:
```

The nested `args` on a `FunctionDef` is a `typing.Optional[typing.List[AssignName]]` so we need the new `hasattr`. 

Also I tried profiling this check. Django didn't seem to have that many private imports to activate this, but when running this on `pylint` the time spent is very small compared to `imports.py` as a whole.

DanielNoord commented:
Do you have a code example for this?

In any case, I think that the `hasattr` is not the correct way to solve this or at the very least it is bypassing a bug from `astroid`. The typing you're referring to was recently added by me as it was previously incorrect. The `hasattr` should always be `True` as even is `args` is `None` `hasattr` should be `True`.
If we're initialising `Arguments` instances without an `args` attribute something is going wrong somewhere. It does make sense to me to do a `if node.args and node.args.args`. Does that work as well?

Pierre-Sassoulas reviewed (commented):
This look better and better ! I have some surface comment that I'll publish now even if I don't have time to review everything right now, sorry.

areveny commented:
Incorporated feedback and added mocked unit tests for checking internal/external packages.

> Do you have a code example for this?

The test log says this but I am not sure how to get a hold of this file. I tried downloading the `logging` module of Python3.10.1 directly but couldn't reproduce this crash. 
```
 test_module_location = '/opt/hostedtoolcache/Python/3.10.1/x64/lib/python3.10'
test_module_name = 'logging'
capsys = <_pytest.capture.CaptureFixture object at 0x7fe8b3a958a0>
```

I'm not sure what this `Property` object is to create a minimal code example.
```
E                   File "/home/runner/work/pylint/pylint/pylint/extensions/private_import.py", line 120, in _populate_type_annotations_function
E                     if node.args:  # pylint: disable=too-many-nested-blocks
E                 AttributeError: 'Property' object has no attribute 'args'
E                 
E               assert not 'Traceback (most recent call last):\n  File "/home/runner/work/pylint/pylint/pylint/utils/ast_walker.py", line 72, in ... node.args:  # pylint: disable=too-many-nested-blocks\nAttributeError: \'Property\' object has no attribute \'args\'\n'
```

For now I have switched the check to be on `node.args.args`.

areveny commented on tests/functional/ext/private_import/private_import.py:
```diff
@@ -0,0 +1,50 @@
+
+# Ignore local modules
+from pylint import _private
+from pylint.checkers import _private
```
Renamed module and added a comment explaining the behavior.

DanielNoord commented:
> For now I have switched the check to be on `node.args.args`.

This is the correct way to solve this. Whenever python modules are actually written in `C` we can't infer the `args` of `Arguments` as we can't "read" the signature of the function like we can for python. That's what `node.args.args == None` is essentially telling us. There is a difference between it being `None` and `[]` (just doesn't have arguments).
I guess there is something within `logging` that is written in C.

I'll Pierre finish his review before I'll take a look at this.

Pierre-Sassoulas reviewed (approved):
I only have nitpicky comments at this point. Great job, and yet another great checker from you :)

Since your commit email is a no-reply, I will take the opportunity to ask you semi privately here:

I've been impressed with your tenacity on getting the 'for any all' checker shipped
and now the current one and also by your general contributions to pylint. I was wondering if you'd be interested in having commit rights to pylint and astroid.

This would allow you a great flexibility in terms of reviewing, triaging or simply getting your code in the repository faster.

Let me know if that would be of any interest to you.

areveny commented:
Thanks! I'll wrap the latest suggestions with DanielNoord's change requests.

I would be interested in being a committer. This is my first open source project and I still have much to learn about Pylint and Astroid but I very much want to continue and expand my involvement.

Thanks again, you and DanielNoord have made this possible by making this project so welcoming and easy to contribute to. I'll be looking to your examples for how to improve repo in this new capacity.

DanielNoord reviewed (changes requested):
Sorry @areveny, perhaps I should have looked at this sooner. I think we can make this a little bit easier to follow with a bit of a refactor.

areveny commented:
Thanks for continued engagement. Integrated most of the feedback. I did a refactor for `visit_importfrom` which improves readability.

> Should this be done in the class or on the instance?

I have kept to the class since the staticmethods are being called in other staticmethods.

I removed the silent try/except on `AttributeError` while searching names for typing usage. Also temporarily the current typing searcher is running on all linted files, not just files with private imports. I will revert this before final release but this is surfacing some more edge cases that crash with `AttributeError`. 

Most of these I will work on resolving in the morning. It seems like a type annotation can be a `BinOp` or `Attribute` for example.

This one from before on `logging` doesn't seem to be getting picked up by the new `node.args.args`. I'm a bit lost on this one if it's coming from a C function.
```
E                     if node.args and node.args.args:
E                 AttributeError: 'Property' object has no attribute 'args'
E                 
E               assert not 'Traceback (most recent call last):\n  File "/home/runner/work/pylint/pylint/pylint/utils/ast_walker.py", line 72, in ...ations_function\n    if node.args and node.args.args:\nAttributeError: \'Property\' object has no attribute \'args\'\n'
```

DanielNoord commented:
One error seems to be on this file:
https://github.com/home-assistant/core/blob/dev/homeassistant/core.py
Please see the logs of: `Primer / run on batch one / 3.8 / Linux (pull_request)`.

For the `logging` module: it seems like `astroid.objects.Property` is a subclass of `nodes.FunctionDef`. You'll probably need to add handle `Property` (and its annotation) a bit differently than `FunctionDef`

DanielNoord commented on pylint/checkers/utils.py:
```diff
@@ -1655,6 +1655,11 @@ def is_typing_guard(node: nodes.If) -> bool:
     ) and node.test.as_string().endswith("TYPE_CHECKING")
 
 
+def is_node_in_typing_guarded_import_block(node: nodes.NodeNG) -> bool:
```
? 😄

DanielNoord commented on tests/functional/ext/private_import/private_import.py:
```diff
@@ -0,0 +1,64 @@
+# pylint: disable=unused-import, missing-docstring, reimported, import-error, wrong-import-order
```
```suggestion
"""Tests for import-private-name."""
# pylint: disable=unused-import, missing-function-docstring, reimported, import-error, wrong-import-order
```

DanielNoord commented on tests/functional/ext/private_import/private_import.txt:
```diff
@@ -0,0 +1,11 @@
+import-private-name:5:0:5:24::Imported private module (_world):UNDEFINED
```
I think we could pass `confidence=HIGH` to the `add_message`. Nothing in this checker is based on inference right? It's all just checking variable names.

DanielNoord commented on tests/functional/ext/private_import/private_import.txt:
```diff
@@ -0,0 +1,11 @@
+import-private-name:9:0:9:30::Imported private module (_city):UNDEFINED
+import-private-name:10:0:10:28::Imported private object (_d):UNDEFINED
+import-private-name:12:0:12:13::Imported private module (_house):UNDEFINED
+import-private-name:13:0:13:29::Imported private module (_house, _chair, _stair):UNDEFINED
```
Can we make it so that when there are multiple imports we say: `Imported private moduleS`? Obviously without the capitalisation.

DanielNoord reviewed (changes requested):
Thanks for the refactor! This is much easier to read.

One other thing: could you add this to our `pylintrc` and see what happens? I think it would be a good to edge-case check to see if this checker passes without any warnings on our own codebase. As far as I know we don't import any private stuff anywhere.

Thinking about this a bit more, a good follow-up issues would be a new setting: `internal-module` in which the user can determine a set of modules that are considered internal to the organization. For example, we would set `astroid` there. We should create a follow-up issue after this has been merged.

areveny commented:
Most of the feedback is incorporated. I also restored not checking for typing usages if there are no private imports in a file.

Items pending:

I'll revisit refactoring the typing guard check util after this PR.

Let me know if you want the static methods on the class changed to instance methods.

Finally, I was expecting the `stdlib` test to fail because I didn't add logic for handling `Property`. I'm glad it passed but the underlying issue might still be there. [Here](https://github.com/PyCQA/pylint/runs/4790957311?check_suite_focus=true) is that error again. This test might be flaky so we could try to recreate it but I haven't been able to create a `Property` with no `args` locally. I think it's an Astroid bug.

Here is the check as it stands run on `pylint`:

```
************* Module pylint.testutils.functional.lint_module_output_update
pylint/testutils/functional/lint_module_output_update.py:8:0: C2601: Imported private module (_pytest.config) (import-private-name)
************* Module pylint.checkers.utils
pylint/checkers/utils.py:82:0: C2601: Imported private module (_string) (import-private-name)
************* Module test_functional
tests/test_functional.py:30:0: C2601: Imported private module (_pytest.config) (import-private-name)
tests/test_functional.py:31:0: C2601: Imported private module (_pytest.recwarn) (import-private-name)
************* Module test_import_graph
tests/test_import_graph.py:27:0: C2601: Imported private module (_pytest.fixtures) (import-private-name)
************* Module pylint.testutils.lint_module_test
pylint/testutils/lint_module_test.py:15:0: C2601: Imported private module (_pytest.config) (import-private-name)
```

DanielNoord commented on pylint/extensions/private_import.py:
```diff
@@ -0,0 +1,190 @@
+                args=("module", private_module_imports[0]),
+                confidence=HIGH,
+            )
+        elif not self.same_root_dir(
+            node, node.modname
+        ):  # Only check imported names if the module is external
```
```suggestion
        # Only check imported names if the module is external
        elif not self.same_root_dir(node, node.modname):
```

DanielNoord commented on pylint/extensions/private_import.py:
```diff
@@ -0,0 +1,190 @@
+
+            names = [n[0] for n in node.names]
+            private_names = self._get_private_imports(names)
+            if (
+                not private_names
+            ):  # No private names, don't have to check against typing annotations
```
```suggestion
            # No private names, don't have to check against typing annotations
            if (not private_names): 
```

DanielNoord commented on pylint/extensions/private_import.py:
```diff
@@ -0,0 +1,190 @@
+
+    def _get_private_imports(
+        self,
+        names: List[str],
```
```suggestion
        names: List[str]
```

Just trying what the line length is here. Can't these be one line?

DanielNoord commented on pylint/extensions/private_import.py:
```diff
@@ -0,0 +1,190 @@
+        self,
+        names: List[str],
+    ) -> List[str]:
+        """Returns the private names from input names by a simple string check"""
```
```suggestion
        """Returns the private names from input names by a simple string check."""
```

I know that `pydocstringformatter` is going to enforce periods at the end a docstring in an upcoming release.

DanielNoord commented on pylint/extensions/private_import.py:
```diff
@@ -0,0 +1,190 @@
+        names: List[str],
+    ) -> List[str]:
+        """Returns the private names from input names by a simple string check"""
+        private_names = [name for name in names if self._name_is_private(name)]
+        return private_names
```
```suggestion
        return [name for name in names if self._name_is_private(name)]
```

DanielNoord commented on pylint/extensions/private_import.py:
```diff
@@ -0,0 +1,190 @@
+        node: nodes.LocalsDictNodeNG,
+        all_used_type_annotations: Set[str],
+    ) -> None:
+        """Adds into the set all_used_type_annotations the names of all names ever used as a type annotation
```
Hmm I just saw that this only checks for "ever".

What about:
```python
import _private
my_var: _private.Thing = _private.Thing()
```

I know the typing doesn't make sense as it is implicit, but wouldn't this be allowed with the current checker?

areveny replied:
Right now this does emit, because in `import X; from X import Y` we only check typing usage on `Y`. This does seem like a frequent pattern and I'll check on `import X` too.

Besides that, a typing usage of a private import masks non-typing usages. It's a trade-off to miss those cases to not search everything for non-type usages.

We could efficiently check for your example since we already have a related node. On an `AnnAssign` annotated with the private type, check if there's a `target` and see that accesses the private module.

areveny replied:
What do you think about the typing/non-typing behavior @DanielNoord? I think besides this it's ready to merge.

DanielNoord replied:
Sorry @areveny! This notification seems to have been lost in my mail, I never saw your reply.

I haven't checked this, but do both of these uses of `Thing()` show up in `for usage_node in node.locals[name]:`?
If so we could do something where `populate_` returns `bool` depending on whether it was used as typing and then do a last check in that `for` loop for whether that specific `usage_node` is typing related. If not > break/return.

I think we should be able to fix this trade-off before merging.

DanielNoord commented on pylint/extensions/private_import.py:
```diff
@@ -0,0 +1,190 @@
+    def _populate_type_annotations_annotation(
+        self,
+        node: Union[nodes.Attribute, nodes.Subscript, nodes.Name],
+        all_used_type_annotations: Set,
```
```suggestion
        all_used_type_annotations: Set[str],
```

DanielNoord reviewed (commented):
Minor formatting and spelling changes.

One question that I just thought about though..

areveny commented:
Phew. I implemented your suggested logic, and if you look at the bottom of the functional tests, we now detect all sorts of cases where an access occurs after a type checking usage. Thanks for encouraging me in this direction, we have a better product because of that.

I also realized there's a case where `from _private import a, b, c` is okay if `a, b, c` are all only used for type checking. This is now supported.

Tragically, the issue with accessing `args` on `Property` cropped up again on (a different) `stdlib` primer [here](https://github.com/PyCQA/pylint/runs/4877951869?check_suite_focus=true). I've added an explicit exclusion for the `Property` node before accessing `args` because tests on `stdlib` would continue to be flaky without it.

It's not ideal to have logic to bypass a potential Astroid bug. I can try to reproduce it locally (probably just by trying to lint the problem file dozens of times) but since it's random it could even be something in the test VM.

DanielNoord commented on tests/functional/ext/private_import/private_import.py:
```diff
@@ -0,0 +1,120 @@
+# However, this depends on annotating a variable and never using it.
+# This would be strange and in function bodies emits `unused-variable`
+import _private_module_unreachable
+my_var8: _private_module_unreachable.Thing8
```
Can we warn here as well, or is that too difficult to do? I don't really like not warning "because another message will also indicate that this is wrong". What if that message is disabled? Then this becomes a false negative.

areveny replied:
I can't think of an straightforward way to close this completely without scanning the whole file. 

We could get closer by counting unassigned annotated variables as invalid uses, which would function if `unused-variable` is disabled.

But this would still fail to the case
```
import public_thing8_maker
import _private_module_unreachable
my_var8: _private_module_unreachable.Thing8
my_var8 = public_thing8_maker.build_thing8()

my_var9 = _private_module_unreachable.Thing8()
```

But I do think we're making progress in these false negatives being increasingly improbable to see in real code.

DanielNoord commented:
> Phew. I implemented your suggested logic, and if you look at the bottom of the functional tests, we now detect all sorts of cases where an access occurs after a type checking usage. Thanks for encouraging me in this direction, we have a better product because of that.

👍 This looks very good indeed. I really hope we're catching all false negatives and positives from the start here 😄 

> I also realized there's a case where `from _private import a, b, c` is okay if `a, b, c` are all only used for type checking. This is now supported.

👍 

> Tragically, the issue with accessing `args` on `Property` cropped up again on the `stdlib` primer [here](https://github.com/PyCQA/pylint/runs/4877951869?check_suite_focus=true). I've added an explicit exclusion for the `Property` node before accessing `args` because tests on `stdlib` would continue to be flaky without it.

I think I found the root cause of this issue. If you want, you can make an `astroid` PR for this as you were the one who identified it. Otherwise I can see if I can find some time to do so.
The issue is in `infer_property` in the `builtin` brain, which also explains why the `stdlib` primer keeps crashing. See:
https://github.com/PyCQA/astroid/blob/9363c34934f94124f4867caf1bdf8f6755201ccd/astroid/brain/brain_builtin_inference.py#L550
`args` on `nodes.FunctionDef` (of which `Property` is a subclass) is defined in the `postinit`. If you search for `Property(` you'll see that there are two places where we instantiate `Properties`. On one we do call `postinit`, but in `infer_property` we don't. This is a bug. We should be instantiating the `Property`, calling `postinit` and then returning the node.
This is an easy fix, I only wonder how this can be tested but I don't think that should be particularly difficult. You can probably just copy the code that is currently crashing and add an assert for the `args` attribute.

> I can try to reproduce it locally (probably just by trying to lint the problem file dozens of times) but since it's random it could even be something in the test VM. With the exclusion the checker sound be safe and I feel okay merging with an issue created to revisit the root cause.

Now that I have identified the issue I would like us to at least open the `astroid` PR/issue so we can add a `TODO` and link to the issue.

areveny commented:
Removed the explicit check on `Property`, disallow the situation where an import is used as a type annotation on a variable that never gets used, e.g.

```
from package import _privateType
a: _privateType
```

Dependent on https://github.com/PyCQA/astroid/pull/1366.

Pierre-Sassoulas commented:
> Dependent on PyCQA/astroid#1366.

We can release a 2.9.x version if required, let me know.

areveny commented:
I'm fine either way. This PR can wait until the next Astroid release.

DanielNoord commented:
@areveny `astroid` has been updated! 😄

areveny commented:
Awesome! Do you think this is ready to merge? I think we've reached a good point in terms of reducing false positives.

I'll merge upstream and run the CI shortly. I just started a new job so I've been busy but will take a look at what else needs to be done for 2.13.

Pierre-Sassoulas commented on doc/whatsnew/2.13.rst:
```diff
@@ -61,6 +61,10 @@ Removed checkers
 Extensions
 ==========
+* New extension ``import-private-name``: indicate imports of external private packages
+  and objects (prefixed with ``_``).
```
```suggestion
  and objects (prefixed with ``_``). It can be loaded using ``load-plugins=pylint.extensions.private_import``.
```

Pierre-Sassoulas commented on ChangeLog:
```diff
@@ -10,6 +10,11 @@ Release date: TBA
 
 
+* New extension ``import-private-name``: indicate imports of external private packages
+  and objects (prefixed with ``_``).
```
```suggestion
  and objects (prefixed with ``_``). It can be loaded using ``load-plugins=pylint.extensions.private_import``.
```

Pierre-Sassoulas reviewed (approved):
>  Do you think this is ready to merge?

Yes :)

DanielNoord reviewed (approved):
What was the thing we were waiting for in `astroid`?

Pierre-Sassoulas commented:
> What was the thing we were waiting for in astroid?

I think @areveny already applied the changes and tested locally, the tests in CI were failing before the merge from main that brought the astroid upgrade.

Pierre-Sassoulas commented:
Amazing work @areveny ! It's finally merged 🎉

DanielNoord commented:
As a follow up, do we want to turn this on ourselves?

areveny commented:
Much thanks for all the careful feedback everyone!
